sounds/management/commands/gm_worker_processing.py

```diff
@@ -49,93 +90,74 @@ def add_arguments(self, parser):
             default='process_sound',
             help='Register this function (default: process_sound)')
 
-    def write_stdout(self, msg):
-        logger.info("[%d] %s" % (os.getpid(), msg))
-
     def handle(self, *args, **options):
-        # N.B. don't take out the print statements as they're
-        # very very very very very very very very very very
-        # helpful in debugging supervisor+worker+gearman
-        self.write_stdout('Starting worker')
+        logger.info('Starting worker')
         task_name = 'task_%s' % options['queue']
-        self.write_stdout('Task: %s' % task_name)
+        logger.info('Task: %s' % task_name)
         if task_name not in dir(self):
-            self.write_stdout("Wow.. That's crazy! Maybe try an existing queue?")
+            logger.info("Wow.. That's crazy! Maybe try an existing queue?")
             sys.exit(1)
+
         task_func = lambda x, y: getattr(Command, task_name)(self, x, y)
-        self.write_stdout('Initializing gm_worker')
+        logger.info('Initializing gm_worker')
         gm_worker = gearman.GearmanWorker(settings.GEARMAN_JOB_SERVERS)
-        self.write_stdout('Registering task %s, function %s' % (task_name, task_func))
+        logger.info('Registering task %s, function %s' % (task_name, task_func))
         gm_worker.register_task(options['queue'], task_func)
-        self.write_stdout('Starting work')
+        logger.info('Starting work')
         gm_worker.work()
-        self.write_stdout('Ended work')
+        logger.info('Ended work')
 
     def task_analyze_sound(self, gearman_worker, gearman_job):
-        return self.task_process_x(gearman_worker, gearman_job, analyze)
+        tmp_directory = None
+        job_data = json.loads(gearman_job.data)
+        sound_id = job_data['sound_id']
+        set_timeout_alarm(settings.WORKER_TIMEOUT, 'Analysis of sound %s timed out' % sound_id)
 
-    def task_process_sound(self, gearman_worker, gearman_job):
-        return self.task_process_x(gearman_worker, gearman_job, process)
+        logger.info("---- Starting analysis of sound with id %s ----" % sound_id)
+        try:
+            check_if_free_space()
+            tmp_directory = tempfile.mkdtemp(prefix='analysis_%s_' % sound_id)
+            result = FreesoundAudioAnalyzer(sound_id=sound_id, tmp_directory=tmp_directory).analyze()
+            if result:
+                logger.info("Successfully analyzed sound %s" % sound_id)
+            else:
+                log_error("Failed analyizing sound %s" % sound_id)
+
+        except WorkerException as e:
+            log_error('%s' % e)
+
+        except Exception as e:
+            log_error('Unexpected error while analyzing sound %s' % e)
 
-    def task_process_x(self, gearman_worker, gearman_job, func):
+        cancel_timeout_alarm()
```

Comment:
In order to avoid passing `None` to `cleanup_tmp_directory`, perhaps we could use try/finally here?
It might involve making the out of disk space exception its own exception instead of `WorkerException`. Alternatively, something like a temp dir context manager might work: https://stackoverflow.com/questions/6884991/how-to-delete-dir-created-by-python-tempfile-mkdtemp
What I"m not sure about here is if mkdtemp can return some kind of exception (it's likely), in which case we probably still need to catch it.

Reply:
After 613a4c9b304d1d4a810854e58711d4854a90afe1 we use a context manager which simplified the cleaning significantly